src/java/org/apache/cassandra/index/sai/disk/format/Version.java

```diff
@@ -0,0 +1,131 @@
+
+import java.util.List;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.Lists;
+
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.disk.v1.V1OnDiskFormat;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Format version of indexing component, denoted as [major][minor]. Same forward-compatibility rules apply as to
+ * {@link org.apache.cassandra.io.sstable.format.Version}.
+ */
+public class Version
+{
+    // Current version
+    public static final Version BA = new Version("ba", V1OnDiskFormat.instance, (c, i) -> defaultFileNameFormat(c, i, "ba"));
+
+    // These should be added in reverse order so that the latest version is used first. Version matching tests
+    // are more likely to match the latest version, so we want to test that one first.
+    public static final List<Version> ALL = Lists.newArrayList(BA);
+
+    public static final Version EARLIEST = BA;
+    // The latest version can be configured to be an earlier version to support partial upgrades that don't
+    // write newer versions of the on-disk formats.
+    public static final Version LATEST = parse(System.getProperty("cassandra.sai.latest.version", "ba"));
+
+    private final String version;
+    private final OnDiskFormat onDiskFormat;
+    private final FileNameFormatter fileNameFormatter;
+
+    private Version(String version, OnDiskFormat onDiskFormat, FileNameFormatter fileNameFormatter)
+    {
+        this.version = version;
+        this.onDiskFormat = onDiskFormat;
+        this.fileNameFormatter = fileNameFormatter;
+    }
+
+    public static Version parse(String input)
+    {
+        checkArgument(input != null);
+        checkArgument(input.length() == 2);
+        if (input.equals(BA.version))
+            return BA;
+        throw new IllegalArgumentException();
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hashCode(version);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Version other = (Version)o;
+        return Objects.equal(version, other.version);
+    }
+
+    @Override
+    public String toString()
+    {
+        return version;
+    }
+
+    public boolean onOrAfter(Version other)
+    {
+        return version.compareTo(other.version) >= 0;
+    }
+
+    public OnDiskFormat onDiskFormat()
+    {
+        return onDiskFormat;
+    }
+
+    public FileNameFormatter fileNameFormatter()
+    {
+        return fileNameFormatter;
+    }
+
+    public interface FileNameFormatter
+    {
+        String format(IndexComponent indexComponent, IndexContext indexContext);
+    }
+
+    //
+    // SAI default filename formatter. This is the current SAI on-disk filename format
+    //
+    // Format: <sstable descriptor>-SAI+<version>(+<index name>)+<component name>.db
+    //
```

Comment:
I'd consider making this a JavaDoc comment for `defaultFileNameFormat()` and briefly mentioning why we wouldn't have an index name (for an SSTable-global file, I'm guessing)